_posts/2022-1-19-quantization-in-practice.md

```diff
@@ -0,0 +1,199 @@
+---
+layout: blog_detail
+title: 'Quantization in Practice'
+author: Suraj Subramanian, Jerry Zhang
+featured-img: ''
+---
+
+There are a few different ways to quantize your model with PyTorch. In this blog post, we'll take a look at how each technique looks like in practice. I will use a non-standard model that is not traceable, to paint an accurate picture of how much effort is really needed when quantizing your model.
+
+<div class="text-center">
+  <img src="/assets/images/quantization_gif.gif" width="60%">
+</div>
+
+## A quick introduction to quantization
+
+> If someone asks you what time it is, you don't respond "10:14:34:430705", but you might say "a quarter past 10".
+
+Quantization has roots in information compression; in deep networks it refers to reducing the numerical precision of its weights and/or activations. Overparameterized DNNs have more degrees of freedom and this makes them good candidates for information compression. When you quantize a model, two things generally happen - the model gets smaller and runs with better efficiency. Processing 8-bit numbers is faster than 32-bit numbers, and a smaller model has lower memory footprint and power consumption.
+
+At the heart of it all is a **mapping function**, a linear projection from floating-point to integer space: $Q(r) = round(r/S + Z)$
+
+To reconvert to floating point space, the inverse function is given by $\tilde r = (Q(r) - Z) \cdot S$. $\tilde r \neq r$, and their difference constitutes the **quantization error**.
+
+The scaling factor $S$ is simply the ratio of the input range to the output range: $S = \frac{\beta - \alpha}{\beta_q - \alpha_q}$
+where [$\alpha, \beta$] is the clipping range of the input, i.e. the boundaries of permissible inputs. [$\alpha_q, \beta_q$] is the range in quantized output space that it is mapped to. For 8-bit quantization, the output range $\beta_q - \alpha_q <= (2^8 - 1) $.
+
+The process of choosing the appropriate input range is known as **calibration**; commonly used methods are MinMax and Entropy. 
```

Comment:
should we mention observers here? calibration is just a step that runs some sample data through the model and also through the inserted observers, so that the values can be recorded and used to calculate quantization parameters